### DISCUSSION TITLE
Transition among state machines

### DISCUSSION THREAD

**Discussion by haocheng6** [statelyai/xstate#2115] [Q&A]
I have several state machines of the same type, for example, A, B, and C. The user can use any of these machines at any time. For example, the user can first use state machine A, then transition to use B, and then go back to A and transition to C, and so on. How should I implement this? The following two approaches seems close, but I need some help.

1. It seems that I can create a bigger state machine that contains A, B, and C, but in this way I need to repeat the same machine for three times in the definition of the bigger state machine. Is there a way that I can reuse the definition of the sub-machine?

2. I am not sure whether this can be implemented with invoked machines because every time a machine is invoked, it is created from scratch without remembering old states, but what I want is to always have three sub-machines and use one of them at some point of time. I don't want any of the sub-machine to be deleted and created again later.

**Comment by Andarist** [comment]
> Is there a way that I can reuse the definition of the sub-machine?

Declare in a variable and use it in three places.

> I don't want any of the sub-machine to be deleted and created again later.

It would be achievable using `invoke` but you would have to store the last known state of a child in your parent and when reinvoking it pass it to it. It probably ain't worth it in your case.

---

As to the original problem. It's rather hard to give any solid advice here beyond what I've written as the problem is abstract and it would be best if you could share more details about it - what are A, B and C, why they are same but yet not the same and how they should interact with each other.
